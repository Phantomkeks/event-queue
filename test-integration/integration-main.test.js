"use strict";

const cds = require("@sap/cds");

cds.test(__dirname + "/_env");

describe("integration-main", () => {
  let context, tx, dbCounts;

  beforeAll(async () => {
    const db = await cds.connect.to("db");
    db.before("*", (cdsContext) => {
      if (dbCounts[cdsContext.event]) {
        dbCounts[cdsContext.event] = dbCounts[cdsContext.event] + 1;
      } else {
        dbCounts[cdsContext.event] = 1;
      }
    });
  });

  beforeEach(() => {
    context = new cds.EventContext({});
    tx = cds.tx(context);
    dbCounts = {};
  });

  afterEach(async () => {
    await tx.rollback();
  });

  afterAll(async () => {
    await cds.disconnect();
    await cds.shutdown();
  });

  it("Test Setup", async () => {
    const eventQueueEntry = await tx.run(SELECT.from("sap.core.EventQueue"));
    await tx.rollback();
    expect(eventQueueEntry).toHaveLength(0);
    expect(dbCounts).toMatchInlineSnapshot(`
      {
        "BEGIN": 1,
        "READ": 1,
        "ROLLBACK": 1,
      }
    `);
  });
});
