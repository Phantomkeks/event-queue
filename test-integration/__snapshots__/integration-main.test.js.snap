// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration-main empty queue - nothing to do 1`] = `
{
  "BEGIN": 1,
  "COMMIT": 1,
  "READ": 1,
}
`;

exports[`integration-main end-to-end insert entry, entry should be automatically processed 1`] = `
{
  "BEGIN": 6,
  "COMMIT": 4,
  "CREATE": 1,
  "READ": 4,
  "ROLLBACK": 1,
  "UPDATE": 3,
}
`;

exports[`integration-main if checkEventAndGeneratePayload methods throws an error --> entry should not be processed + status should be error 1`] = `
[
  [
    "Unexpected error during event processing - setting queue entry to error",
    {
      "additionalMessageProperties": {
        "eventSubType": "Task",
        "eventType": "Notifications",
        "queueEntriesIds": [
          "dbaa22d5-41db-4ff3-bdd8-e0bb19b217cf",
        ],
      },
      "error": [Error: error during processing],
    },
  ],
]
`;

exports[`integration-main if checkEventAndGeneratePayload methods throws an error --> entry should not be processed + status should be error 2`] = `
{
  "BEGIN": 3,
  "COMMIT": 2,
  "READ": 2,
  "UPDATE": 2,
}
`;

exports[`integration-main if cluster methods throws an error --> entry should not be processed + status should be error 1`] = `
[
  [
    "Error during clustering of events - setting all queue entries to error",
    {
      "additionalMessageProperties": {
        "eventSubType": "Task",
        "eventType": "Notifications",
      },
      "error": [Error: error during processing],
    },
  ],
]
`;

exports[`integration-main if cluster methods throws an error --> entry should not be processed + status should be error 2`] = `
{
  "BEGIN": 4,
  "COMMIT": 2,
  "READ": 3,
  "ROLLBACK": 1,
  "UPDATE": 2,
}
`;

exports[`integration-main if modifyQueueEntry methods throws an error --> entry should not be processed + status should be error 1`] = `
[
  [
    "Unexpected error during event processing - setting queue entry to error",
    {
      "additionalMessageProperties": {
        "eventSubType": "Task",
        "eventType": "Notifications",
        "queueEntriesIds": [
          "dbaa22d5-41db-4ff3-bdd8-e0bb19b217cf",
        ],
      },
      "error": [Error: error during processing],
    },
  ],
]
`;

exports[`integration-main if modifyQueueEntry methods throws an error --> entry should not be processed + status should be error 2`] = `
{
  "BEGIN": 3,
  "COMMIT": 2,
  "READ": 2,
  "UPDATE": 2,
}
`;

exports[`integration-main if process event returns an error --> tx should be rolled backed 1`] = `
{
  "BEGIN": 6,
  "COMMIT": 3,
  "READ": 5,
  "ROLLBACK": 2,
  "UPDATE": 3,
}
`;

exports[`integration-main if process event throws an error --> tx should be rolled backed 1`] = `
{
  "BEGIN": 6,
  "COMMIT": 3,
  "READ": 5,
  "ROLLBACK": 2,
  "UPDATE": 3,
}
`;

exports[`integration-main insert one entry and process 1`] = `
{
  "BEGIN": 5,
  "COMMIT": 3,
  "READ": 4,
  "ROLLBACK": 1,
  "UPDATE": 3,
}
`;
